### ISSUE TITLE
Why deepCopy is not public?

### ISSUE THREAD

**Issue by kramer** [google/gson#760]
Is there a reason why the `deepCopy` methods are not public? I see that they were public for some time in an earlier revision then was made "default" access level but I couldn't find the reason.

**Comment by inder123** [comment]
It was never made public in a released version. No specific reason, we just didn't see it useful enough to bloat our API. What are you trying to do that you need it?

**Comment by kramer** [comment]
> It was never made public in a released version.

Fair enough, I was only checking the file history and didn't notice that, sorry.

> What are you trying to do that you need it?

To be exact for [this](https://github.com/searchbox-io/Jest/blob/master/jest-common/src/main/java/io/searchbox/client/JestResult.java#L203) but a shorter version is:

```java
class Demo {
  private JsonObject serverResponse;

  public JsonObject getResultsFromResponse() {
    JsonObject results = serverResponse.getAsJsonObject("results");
    results.add("SOME_DEFAULT_DATA", "");
    return results;
  }
}
```

In this code piece calling `getResultsFromResponse` will change content of the class variable `serverResponse` but actually what I needed was to add the default data field only to the object returned by that method. So what I wanted to do is create a copy of `results` then add the default field to it and return so that `serverResponse` would not be affected - but this is not possible without access to `deepCopy`.

Let me know if I overlooked something or if this is possible to do without copying.

**Comment by JakeWharton** [comment]
We could support deep copy by exposing clone() with a covariant return type
on these types. We do this in OkHttp and Retrofit to great effect (not that
I'm biased towards those libraries or anything).

**Comment by swankjesse** [comment]
Two things:
 * I think it’s conventional for mutable value objects like this to have `clone()` return a shallow copy. So we’d probably want to name the method `deepCopy()`, and possibly also have `clone()` that does a shallow copy. (Reason being, the default `Object.clone()` is a shallow copy).
 * It’s possible to create a mutable `JsonPrimitive` by creating it with an `AtomicInteger`. If we expose a mechanism to deep copy, we should probably convert numbers to `LazilyParsedNumber` instances.

**Comment by swankjesse** [comment]
And for what it’s worth, a shallow copy is sufficient to satisfy @kramer’s request.

**Comment by kramer** [comment]
> And for what it’s worth, a shallow copy is sufficient to satisfy @kramer’s request.

May be true for the exact code I posted but are you sure that a shallow copy is still sufficient if the modification (the field addition) was in a deeper level (e.g.: `results.getAsJsonObject("failed").add(...)` - here the `failed` object will still be shared between the class var and the returned copy if I'm not mistaken) ?

**Comment by swankjesse** [comment]
Yep, in that case you’ll need a deep copy.

**Comment by sinaci** [comment]
I came across this thread when I encountered the fact that I cannot extend JsonElement because of the default access level of the deepCopy. Is there any alternative for me? I want to implement my own class which extends JsonElement.

**Comment by SSeanSS** [comment]
Hi @kramer, so far have you found any solution to solve the issue? I'm right facing the same issue as you did..

**Comment by balajeetm** [comment]
Hi,

I'm looking at ways to merge Jsons the way its available in lodash.
I couldn't find it with Gson. Is it available?
If not, I'm planning to write one. DeepCopy would be helpful here.
Can this is be made public?

Regards,

**Comment by inder123** [comment]
No specific reason to hide deepCopy (except to have a smaller API set). 
For now, you can copy over the code?

**Comment by SSeanSS** [comment]
@inder123 now i did the deepCopy in this way 
`
/**
     * This method makes a "deep copy" of JsonElement it is given.
     */
    public static JsonElement deepCopy(JsonElement element) {
        try {
            String elementString = element.toString();
            return new JsonParser().parse(elementString);
        }
        catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }
`

**Comment by khoanguyen123** [comment]
+1 for making deepCopy() public. My code below would alter the source, which I'd like to prevent:

    private JsonElement removeAttribute(String attribute, JsonObject source) {
        source.remove(attribute);
        return source;
    }

**Comment by artfiedler** [comment]
+1 deepCopy() should be public, I would use it when merging two JsonObject's into one when the JsonObjects are settings and default settings... like so...

```
public static JsonObject merge(String overrideJson, JsonObject defaultObj) {
    return mergeInto((JsonObject)new JsonParser().parse(overrideJson), defaultObj);
}
public static JsonObject merge(JsonObject overrideObj, JsonObject defaultObj) {
    return mergeOverride((JsonObject)deepClone(defaultObj), overrideObj);
}
public static JsonObject mergeOverride(JsonObject targetObj, JsonObject overrideObj) {
    for (Map.Entry<String, JsonElement> entry : overrideObj.entrySet())
            targetObj.add(entry.getKey(), deepClone(entry.getValue()));
    return targetObj;
}
public static JsonObject mergeInto(JsonObject targetObj, JsonObject defaultObj) {
    for (Map.Entry<String, JsonElement> entry : defaultObj.entrySet()) {
        if (targetObj.has(entry.getKey()) == false)
            targetObj.add(entry.getKey(), deepClone(entry.getValue()));
    }
    return targetObj;
}
public static JsonElement deepClone(JsonElement el){
    if (el.isJsonPrimitive() || el.isJsonNull())
        return el;
    if (el.isJsonArray()) {
        JsonArray array = new JsonArray();
        for(JsonElement arrayEl: el.getAsJsonArray())
            array.add(deepClone(arrayEl));
        return array;
    }
    if(el.isJsonObject()) {
        JsonObject obj = new JsonObject();
        for (Map.Entry<String, JsonElement> entry : el.getAsJsonObject().entrySet()) {
            obj.add(entry.getKey(), deepClone(entry.getValue()));
        }
        return obj;
    }
    throw new IllegalArgumentException("JsonElement type " + el.getClass().getName());
}
```

**Comment by mykhaylo-** [comment]
+1 to make the _deepCopy_ public. This feature is naturally expected to be provided by the gson library. There are lot of cases where you just want to take some your JsonObject as a source (or template) for your further manipulations(adding/removing properties, etc) but keeping the original untouched. with _deepCopy_ you don't need to reinvent a wheel in every of your projects. you simply use it out of the box.

**Comment by bohdaq** [comment]
Guys please make deepCopy public...

**Comment by GeniusWiki** [comment]
yes, it needs a way to clone.  I prefer just implements Cloneable  interface and just make it deepCopy.

**Comment by inder123** [comment]
I am open to it. Please send a pull request. Thanks.

**Comment by CateyesLin** [comment]
+1
I need JsonElement make deepCopy public for doing merge/extend 2 JsonObject.

**Comment by neowang813** [comment]
+1

**Comment by toant13** [comment]
+1

**Comment by adaviding** [comment]
+1   make it public

**Comment by thinkingmedia** [comment]
Do it already...

**Comment by jakekdodd** [comment]
+1

**Comment by artfiedler** [comment]
+42

**Comment by tianhao** [comment]
+1 need public

**Comment by Ankhwatcher** [comment]
+1 My life would be easier if this were public.

**Comment by inder123** [comment]
https://github.com/google/gson/pull/1091

**Comment by djdance** [comment]
very useful function, will wait for it

**Comment by faissaloo** [comment]
It's public now, please close.

**Comment by Marcono1234** [comment]
@eamonnmcmanus, this was resolved by #1091.
